lib/primer/classify/cache.rb

```diff
@@ -9,107 +9,124 @@ module Primer
   class Classify
     # :nodoc:
     class Cache
-      # rubocop:disable Style/MutableConstant
-      LOOKUP = {}
-      # rubocop:enable Style/MutableConstant
-
-      class <<self
-        def read(memo, key, val, breakpoint)
-          value = LOOKUP.dig(breakpoint, key, val)
-          memo[:classes] << value if value
-        end
+      include Singleton
```

Comment:
I'm not really familiar with Singletons in Rails, since most apps run multiple instances, will each instance have a different cache?

Reply:
That's a good question. I believe we run Unicorn in production, which preloads the app and then forks into child processes. Any time the program writes to a shared place in memory (i.e. changes a variable), the memory is copied, then mutated (this is known as copy-on-write, or CoW). Since it's very likely the singleton instance will get mutated _after_ forking, each process will make a copy of the original `@lookup` hash. If my mental model is correct, that's the same as what happened when we used the `LOOKUP` constant.

Reply:
We could make this explicit by implementing our own singleton logic that could store the instance per-thread (eg. per-process) in `Thread.current`. I actually tried that earlier this week and found that it negatively affected performance.